src/Microsoft.AspNetCore.Sockets.Abstractions/DefaultConnectionContext.cs

```diff
@@ -22,16 +22,26 @@ public class DefaultConnectionContext : ConnectionContext,
                                             IConnectionHeartbeatFeature,
                                             ITransferFormatFeature
     {
-        private List<(Action<object> handler, object state)> _heartbeatHandlers = new List<(Action<object> handler, object state)>();
+        private object _heartBeatLock = new object();
+        private List<(Action<object> handler, object state)> _heartbeatHandlers;
+        private PipeOptions _transportPipeOptions;
+        private PipeOptions _applicationPipeOptions;
+        private IDuplexPipe _transportPipe;
+        private IDuplexPipe _applicationPipe;
 
         // This tcs exists so that multiple calls to DisposeAsync all wait asynchronously
         // on the same task
         private TaskCompletionSource<object> _disposeTcs = new TaskCompletionSource<object>();
 
-        public DefaultConnectionContext(string id, IDuplexPipe transport, IDuplexPipe application)
+        public DefaultConnectionContext(string id)
```

Comment:
I'm not a fan of these changes. I think you should just set the pipes on the outside. The options shouldn't be part of this class at all.

Reply:
Do we care about people using this layer having to do that themselves?

Reply:
I think it's OK to require that callers have to set the pipe themselves. User applications (Endpoints) will get a fully configured `ConnectionContext`, so we're the only consumers of this during initialization.

Reply:
We can keep the other ctor to avoid changing all of the callers. We're just adding a new optional overload.